metricproviders/datadog/datadog.go

```diff
@@ -0,0 +1,197 @@
+package datadog
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	"github.com/argoproj/argo-rollouts/utils/evaluate"
+	metricutil "github.com/argoproj/argo-rollouts/utils/metric"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/jsonpath"
+)
+
+var unixNow = func() int64 { return time.Now().Unix() }
+
+const (
+	//ProviderType indicates the provider is datadog
+	ProviderType            = "Datadog"
+	DatadogTokensSecretName = "datadog"
+)
+
+// Provider contains all the required components to run a Datadog query
+// Implements the Provider Interface
+type Provider struct {
+	logCtx log.Entry
+	apiKey string
+	appKey string
+}
+
+// Type incidates provider is a Datadog provider
+func (p *Provider) Type() string {
+	return ProviderType
+}
+
+func (p *Provider) Run(run *v1alpha1.AnalysisRun, metric v1alpha1.Metric) v1alpha1.Measurement {
+	startTime := metav1.Now()
+
+	// Measurement to pass back
+	measurement := v1alpha1.Measurement{
+		StartedAt: &startTime,
+	}
+
+	endpoint := "https://api.datadoghq.com/api/v1/query"
+	if metric.Provider.Datadog.Address != "" {
+		endpoint = metric.Provider.Datadog.Address + "/api/v1/query"
+	}
+
+	url, _ := url.Parse(endpoint)
+
+	now := unixNow()
+	var interval int64 = 300
+	if metric.Provider.Datadog.Interval != "" {
+		expDuration, err := metric.Provider.Datadog.Interval.Duration()
+		if err != nil {
+			return metricutil.MarkMeasurementError(measurement, err)
+		}
+		// Convert to seconds as DataDog expects unix timestamp
+		interval = int64(expDuration.Seconds())
+	}
+
+	q := url.Query()
+	q.Set("query", metric.Provider.Datadog.Query)
+	q.Set("from", strconv.FormatInt(now-interval, 10))
+	q.Set("to", strconv.FormatInt(now, 10))
+	url.RawQuery = q.Encode()
+
+	request := &http.Request{Method: "GET"}
+	request.URL = url
+	request.Header = make(http.Header)
+	request.Header.Set("Content-Type", "application/json")
+	request.Header.Set("DD-API-KEY", p.apiKey)
+	request.Header.Set("DD-APPLICATION-KEY", p.appKey)
+
+	// Send Request
+	httpClient := &http.Client{
+		Timeout: time.Duration(10) * time.Second,
+	}
+	response, err := httpClient.Do(request)
+
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	value, status, err := p.parseResponse(metric, response)
+	if err != nil {
+		return metricutil.MarkMeasurementError(measurement, err)
+	}
+
+	measurement.Value = value
+	measurement.Phase = status
+	finishedTime := metav1.Now()
+	measurement.FinishedAt = &finishedTime
+
+	return measurement
+}
+
+func (p *Provider) parseResponse(metric v1alpha1.Metric, response *http.Response) (string, v1alpha1.AnalysisPhase, error) {
+	var data interface{}
+
+	jsonParser := jsonpath.New("metrics")
+	jsonParser.Parse("{.series[0].pointlist[-1:][1]}")
```

Comment:
Instead of using jsonParser to access the float array we are interested in, could we just unmarshal the body into a data type which we expect it to be in? e.g.:

```
type datadogResponse struct {
	Series []struct {
		Pointlist [][]float64 `json:"pointlist"`
	}
}
```

I think theres a lot of overhead of using jsonpath/reflection approach currently.

Reply:
Refactored this so its no longer use jsonpath